vault/ssorolecredentialsprovider.go

```diff
@@ -0,0 +1,272 @@
+package vault
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/99designs/keyring"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/service/sso"
+	"github.com/aws/aws-sdk-go/service/ssooidc"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/skratchdot/open-golang/open"
+)
+
+const (
+	ssoClientName         = "aws-vault"
+	ssoClientType         = "public"
+	oAuthTokenGrantType   = "urn:ietf:params:oauth:grant-type:device_code"
+	authorizationTemplate = `
+Attempting to automatically open the SSO authorization page in your default
+browser. If the browser does not open or you wish to use a different device to
+authorize this request, open the following URL:
+
+%s
+(Use Ctrl-C to abort)
+
+`
+)
+
+// CachedSSORoleCredentialsProvider uses the keyring to cache SSO Role sessions.
+type CachedSSORoleCredentialsProvider struct {
```

Comment:
Hmm I was planning a sessions refactor, but it may need to encompass this. Mind if I ping you if I get to this?